util/src/main/java/tech/pegasys/artemis/util/hashtree/HashTreeUtil.java

```diff
@@ -15,116 +15,98 @@
 
-import com.google.common.primitives.UnsignedLong;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
 import net.consensys.cava.bytes.Bytes;
 import net.consensys.cava.bytes.Bytes32;
 import net.consensys.cava.crypto.Hash;
-import net.consensys.cava.ssz.SSZ;
 
 /** This class is a collection of tree hash root convenience methods */
 public final class HashTreeUtil {
 
-  /**
-   * Calculate the hash tree root of the provided value
-   *
-   * @param value
-   */
-  public static Bytes32 hash_tree_root(Bytes value) {
-    return SSZ.hashTreeRoot(value);
+  public static enum SSZTypes {BASIC, CONTAINER, TUPLE_BASIC, TUPLE_CONTAINER, LIST_BASIC, LIST_CONTAINER};
+  public static final int BYTES_PER_CHUNK = 32;
+
+  private static Bytes32 hash_tree_root(SSZTypes sszType, Bytes... bytes) {
+    // NOT YET IMPLEMENTED
+    // TODO - My plan is to wrap all the hash_tree_root calculation in this class so our types aren't as complex.
+    return Bytes32.ZERO;
   }
 
   /**
-   * Calculate the hash tree root of the provided value
+   * Create the hash tree root of a set of values of basic SSZ types or tuples of basic types. Basic
+   * SSZ types are uintN, bool, and byte. bytesN (i.e. Bytes32) is a tuple of basic types. NOTE:
+   * Bytes (and not Bytes32, Bytes48 etc.) IS NOT a basic type or a tuple of basic types.
    *
-   * @param value
+   * @param bytes One Bytes value or a list of homogeneous Bytes values.
+   * @return The SSZ tree root hash of the values.
+   * @see <a
+   *     href="https://github.com/ethereum/eth2.0-specs/blob/v0.5.1/specs/simple-serialize.md">SSZ
+   *     Spec v0.5.1</a>
    */
-  public static Bytes32 hash_tree_root(UnsignedLong val) {
-    return SSZ.hashTreeRoot(
-        SSZ.encode(
-            writer -> {
-              writer.writeUInt64(val.longValue());
-            }));
+  public static Bytes32 hash_tree_root_basic_type(Bytes... bytes) {
+    return merkleize(pack(bytes));
   }
 
   /**
-   * Calculate the hash tree root of the list of validators provided
+   * Create the hash tree root of a list of values of basic SSZ types. This is only to be used for
+   * SSZ lists and not SSZ tuples. See the "see also" for more info.
    *
-   * @param list
+   * @param bytes A list of homogeneous Bytes values representing basic SSZ types.
+   * @return The SSZ tree root hash of the list of values.
+   * @see <a
+   *     href="https://github.com/ethereum/eth2.0-specs/blob/v0.5.1/specs/simple-serialize.md">SSZ
+   *     Spec v0.5.1</a>
    */
-  public static Bytes32 hash_tree_root(List<Bytes> list) {
-    return hash_tree_root(
-        SSZ.encode(
-            writer -> {
-              writer.writeBytesList(list);
-            }));
+  public static Bytes32 hash_tree_root_list_of_basic_type(List<? extends Bytes> bytes, int length) {
+    return mix_in_length(hash_tree_root_basic_type(bytes.toArray(new Bytes[0])), length);
```

Comment:
I don't understand why you have `new Bytes[0]` here

Reply:
new Bytes[0] is just a type hint so that toArray creates an array of Bytes and not Object. https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#toArray-T:A-